### PR TITLE
592 derive vars dtm() date imp

### PR DESCRIPTION
**Pull request by bms63**
Thank you for your Pull Request! We have developed this task checklist from the [Development Process Guide](https://roche-gsk.github.io/admiral/articles/development_process.html) to help with the final steps of the process. Completing the below tasks helps to ensure our reviewers can maximize their time on your code as well as making sure the admiral codebase remains robust and consistent.   

Please check off each taskbox as an acknowledgment that you completed the task or check off that it is not relevant to your Pull Request. This checklist is part of the Github Action workflows and the Pull Request will not be merged into the `devel` branch until you have checked off each task.

- [x] Code is formatted according to the [tidyverse style guide](https://style.tidyverse.org/) 
- [x] Updated relevant unit tests or have written new unit tests - See [Unit Test Guide](https://roche-gsk.github.io/admiral/articles/unit_test_guidance.html#writing-unit-tests-in-admiral-)
- [x] If you removed/replaced any function and/or function parameters, did you fully follow the [deprecation guidance](https://roche-gsk.github.io/admiral/articles/programming_strategy.html#deprecation-1)?
- [x] Update to all relevant roxygen headers and examples 
- [x] Run `devtools::documents()` so all `.Rd` files in the `man` folder and the `NAMESPACE` file in the project root are updated appropriately
- [x] Address any updates needed for vignettes and/or templates
- [x] Update `NEWS.md`
- [x] Build admiral site `pkgdown::build_site()` and check that all affected examples are displayed correctly and that all new functions occur on the "[Reference](https://roche-gsk.github.io/admiral/reference/index.html)" page. 
- [x] Address or fix all lintr warnings and errors - `lintr::lint_package()`
- [x] Run `R CMD check` locally and address all errors and warnings - `devtools::check()`
- [x] Link the issue so that it closes after successful merging. 
- [x] Address all merge conflicts and resolve appropriately 
- [x] Pat yourself on the back for a job well done!  Much love to your accomplishment!

### CODE REVIEW

File: R/derive_date_vars.R [hunk 1]
```diff
@@ -69,6 +69,18 @@
 #' cut off date. Only dates which are in the range of possible dates are
 #' considered.
 #'
+#' @param preserve Preserve partial dates when doing date imputation for middle
```
Comment (bundfussr) [inline R/derive_date_vars.R:72]: The `preserve` parameter should affect all imputations not only `date_imputation = "mid"`.

**Review by bundfussr — CHANGES_REQUESTED** [review]
- In the "Breaking Changes" section should be mentioned that the behaviour for `date_imputation = "mid"` has been changed.
- There are some functions which provide the `date_imputation` parameter (`derive_disposition_dt()` and `lstalvdt_source()`). I think they should provide the `preserve` parameter as well.
- I am not sure what we should do for functions which provide no `date_imputation` parameter but impute dates, e.g., `filter_date_source()`, which is called by `derive_param_tte()`. What do you think?

File: R/derive_date_vars.R (hunk 1 above)
Comment (bms63) [reply R/derive_date_vars.R:72]: As I was finishing this up - I was like maybe we will want to use this preserve function more?  Should this be created as a separate issue as this one just focus on "mid" in this issue?

> * In the "Breaking Changes" section should be mentioned that the behaviour for `date_imputation = "mid"` has been changed.
> * There are some functions which provide the `date_imputation` parameter (`derive_disposition_dt()` and `lstalvdt_source()`). I think they should provide the `preserve` parameter as well.
> * I am not sure what we should do for functions which provide no `date_imputation` parameter but impute dates, e.g., `filter_date_source()`, which is called by `derive_param_tte()`. What do you think?

Hey Stefan.  Many thanks for reviewing.  

- I will update the Breaking Changes section.
- Should this use of the preserve parameter be a separate issue as I'm changing multiple functions now?  Nervous to change so much in one Pull Request.  
- I'm not really sure - perhaps something we should bring up to the wider group?

**Comment by bms63** [comment]
> * In the "Breaking Changes" section should be mentioned that the behaviour for `date_imputation = "mid"` has been changed.
> * There are some functions which provide the `date_imputation` parameter (`derive_disposition_dt()` and `lstalvdt_source()`). I think they should provide the `preserve` parameter as well.
> * I am not sure what we should do for functions which provide no `date_imputation` parameter but impute dates, e.g., `filter_date_source()`, which is called by `derive_param_tte()`. What do you think?

Hey Stefan.  Many thanks for reviewing.  

- I will update the Breaking Changes section.
- Should this use of the preserve parameter be a separate issue as I'm changing multiple functions now?  Nervous to change so much in one Pull Request.  
- I'm not really sure - perhaps something we should bring up to the wider group?

File: R/derive_date_vars.R (hunk 1 above)
Comment (bundfussr) [reply R/derive_date_vars.R:72]: As the `preserve` parameter is already implemented in this issue, I see no benefit in moving it into a separate issue.

**Comment by bundfussr** [comment]
If a change requires updating many functions I would do it in one PR rather than using separate ones. Otherwise, `devel` may end up in an "incomplete" state when only one of PR is merged but not the other. Thus I would prefer updating `derive_disposition_dt()` and `lstalvdt_source()` in this PR.

I have put the third item regarding `derive_param_tte()` on the agenda of the Monday meeting.

**Comment by bms63** [comment]
Hey Stefan @bundfussr , I have updated my branch on your suggestions.  I'm having trouble with the cyclometic complexity.  I tried updating the `.lintr` file just to ignore for a time, but hoping we could increase the limit to 20.  Unfortunately, I think the lintr is triggered somewhere else and so getting linr failing.  #560.  

@thomas-neitmann Any chance we could increase the lintr for the cyclometic complexity.

The issue for this "issue" is in the `impute_dtc` function.  Invoking the preserve and date_imputation agrguments can make this function get a bit out of hand with the call to `case_when `and and all the if/else statements. .  Perhaps I should of written a helper function to deal with this preserve and date_imputation stuff before hand but I was thinking this would just be a quick update!  But I think this is getting a bit more complicated. 

Ben

**Comment by bms63** [comment]
@bundfussr  I might of rebuilt the `impute_dtc` function to take out a bunch of the cyclomatic complexity, but haven't fully tested.  Hold off on review.  Thanks!!

**Comment by bms63** [comment]
@bundfussr Can't figure it out.  My mind is getting stuck in weird if/else statement combos.  Hopefully, we can increase cyclomatic complexity to 20 and this will be a none-issue.  Although I don't like how the code is turning out as it is just one big mess!! If you have any ideas on my current Pull Request to clean it/make it more readable please share - I think this is a bit above my skill level at the moment.

Below I have some sample code where I was trying to clean it up and make it more elegant (hopefully) while reducing the cyclomatic complexity, but I'm failing unit tests that involve other arguments later on in the function :(   Here is what I was trying to do.  If you have any ideas on how I could tackle my Pull Request or my idea below, please let me know.  Thanks for your help!!

```
# Specific setup for FIRST/MID/LAST
    # make keywords case-insensitive
    if (date_imputation %in% c("FIRST", "MID", "LAST") && !preserve) {
      d <- list(FIRST = "01", MID = "15", LAST = "31")[[date_imputation]]
      mo <- list(FIRST = "01", MID = "06", LAST = "12")[[date_imputation]]

    } else if (date_imputation %in% c("FIRST", "MID", "LAST") && preserve) {

      d <- list(FIRST = ifelse(n_chr %in% c(4, 7), "01", substr(dtc, 8, 9)),
                MID   = ifelse(n_chr %in% c(4, 7), "15", substr(dtc, 8, 9)),
                LAST  = ifelse(n_chr %in% c(4, 7), "31", substr(dtc, 8, 9))
                )[[date_imputation]]
      mo <- list(FIRST = ifelse(n_chr %in% c(7), substr(dtc, 6, 7), "01"),
                 MID = ifelse(n_chr %in% c(7), substr(dtc, 6, 7), "06"),
                 LAST = ifelse(n_chr %in% c(7), substr(dtc, 6, 7), "12")
                 )[[date_imputation]]
```

**Comment by bundfussr** [comment]
@bms63 , I think the easiest way to reduce the cyclomatic complexity is to put parts of the code in subfunctions, e.g., one subfunction for imputing the date part and another one for imputing the time part.

To make it more readable and maybe reduce the number of cases you could try something like the following.
```
  # extract input components
  components <- str_match(dtc, "(\\d{4}|)(-|)(\\d{2}|-|)(-|)(\\d{2}|)")
  in_year <- components[,2]
  in_month <- components[,4]
  in_day <- components[,6]
  year_miss <- in_year %in% c("","-")
  month_miss <- in_month %in% c("","-")
  day_miss <- in_day %in% c("","-")
  
  # extract requested imputation components
  <as before>

  # derive output components
  out_month <- if_else(month_miss, imp_month, in_month)
  out_day <- if_else(day_miss | month_miss & !preserve, imp_day, out_day)

  ...
```

**Comment by bms63** [comment]
@bundfussr Hey Stefan.  I was able to use the case_when function to take out a lot of the if/else statements.  I think this is ready for your review again.  I do think long term we should refactor this code, but maybe I should focus on other things!?!

**Review by bundfussr — CHANGES_REQUESTED** [review]
According to our decision on Monday the `preserve` parameter should be added to functions which do imputation but do not provide a `date_imputation` parameter. This should affect `derive_param_tte()`, `derive_var_trtedtm()`, and `derive_var_trtsdtm()`.

**Comment by bundfussr** [comment]
I agree that we should refactor the code. I think this PR demonstrates that there is room for improvement.

Either we should do it now or when implementing https://github.com/pharmaverse/admiral/issues/727.

File: R/derive_date_vars.R [hunk 2]
```diff
@@ -179,22 +193,55 @@ impute_dtc <- function(dtc,
...
+      n_chr == 4 & date_imputation != "MID" ~ paste0(dtc, "-", mo, "-", d),
+      n_chr == 4 & date_imputation == "MID" ~ paste0(dtc, "-", "06", "-", "30")
     )
 
-    if (date_imputation == "LAST") {
-      imputed_date <- case_when(
-        nchar(imputed_date) > 0 & n_chr < 10 ~ as.character(ceiling_date(as.Date(imputed_date, format = "%Y-%m-%d"), "month") - days(1)), # nolint
+    # This section is focused on date_imputation wth "LAST" and preserve
+    imputed_date <- case_when(
+       n_chr < 10 & date_imputation == "LAST" & !preserve ~
```
[the remark this answers is not in the document]
Comment (bms63) [reply R/derive_date_vars.R:221]: If I use that if statement then I get the lintr issue with cyclomatic complexity.  Big reason to refactor the the code!

[the remark this answers is not in the document]
Comment (bundfussr) [reply R/derive_date_vars.R:221]: I see. However, the intention of checking the cyclomatic complexity is to make the code more readable. If we can achieve a low cyclomatic complexity only by making the code less readable, I would rather accept the higher cyclomatic complexity.

**Comment by bms63** [comment]
There is at least 9 issues that could be brought into PR! 
image

File: R/derive_date_vars.R (hunk 2 above)
[the remark this answers is not in the document]
Comment (bms63) [reply R/derive_date_vars.R:221]: Then @thomas-neitmann would need to increase the cyclomatic complexity in the lintr file.  #560 it appears we discussed increasing it from 15 to 20 before, but I don't think it was implemented.  Happy to put the if/else statements back in and agree they do make the code more readable.

**Comment by bundfussr** [comment]
I think #753, #727, #701, #700, and #607 do not require an update of the imputation functions.

**Comment by bms63** [comment]
Good point I just search for anything related to `_dt`.  Not the best search, but I think there is more out there.

File: R/derive_date_vars.R [hunk 3]
```diff
@@ -179,22 +193,55 @@ impute_dtc <- function(dtc,
...
+      n_chr == 9 & date_imputation == "LAST" & preserve ~
+        paste0(substr(dtc, 1, 4), "-", 12, "-", substr(dtc, 8, 9)),
+      n_chr == 7 & date_imputation == "LAST" & preserve ~
+        as.character(
+          ceiling_date(
+            as.Date(imputed_date, format = "%Y-%m-%d"), "month") - days(1)),
+      n_chr == 4 & date_imputation == "LAST" & preserve ~
+        as.character(
+          ceiling_date(
+            as.Date(imputed_date, format = "%Y-%m-%d"), "month") - days(1)),
```
[the remark this answers is not in the document]
Comment (bms63) [reply R/derive_date_vars.R]: Simplifed into one

File: R/derive_date_vars.R [hunk 4]
```diff
@@ -179,22 +193,55 @@ impute_dtc <- function(dtc,
...
+      TRUE ~ imputed_date)
+
+
+    imputed_date <- case_when(
+      n_chr == 9 & date_imputation == "FIRST" & preserve ~
+          paste0(substr(dtc, 1, 4), "-", 01, "-", substr(dtc, 8, 9)),
```
[the remark this answers is not in the document]
Comment (bms63) [reply R/derive_date_vars.R:249]: RIght! Thanks for the good catch.  I fixed this one and another one that was similar.

File: R/derive_date_vars.R (hunk 2 above)
[the remark this answers is not in the document]
Comment (bms63) [reply R/derive_date_vars.R]: I've put in all the if/else needed.  I have successfully maxed out the complexity to 19!  whew!

File: NEWS.md [hunk 5]
```diff
@@ -13,8 +19,14 @@ are not recorded in the data. (#589)
 `derive_vars_last_dose()`, and three wrapper functions `derive_var_last_dose_amt()`, 
 `derive_var_last_dose_date()`, and `derive_var_last_dose_grp()` (#385)
 
+## Documentation
+
 - New vignette [Development Process](../articles/development_process.html) and improvements made to contribution vignettes (#765 & #758)
 
+## Breaking Changes
+
+- Behavior of `derive_vars_dtm()` and `derive_vars_dt()` has changed with regards to `date_imputation`.
```
[the remark this answers is not in the document]
Comment (bms63) [reply NEWS.md:28]: K this has been updated and has issue number linked

File: R/derive_date_vars.R [hunk 6]
```diff
@@ -179,22 +193,55 @@ impute_dtc <- function(dtc,
...
       n_chr >= 10 ~ substr(dtc, 1, 10),
+
+      # dates like 2021---14 - use only year part
+      n_chr == 9 & date_imputation == "MID" & !preserve ~
+        paste0(substr(dtc, 1, 4), "-", mo, "-", d),
+
       # dates like 2021---14 - use only year part
-      n_chr == 9 ~ paste0(substr(dtc, 1, 4), "-", mo, "-", d),
+      n_chr == 9 & date_imputation != "MID" & !preserve ~
+        paste0(substr(dtc, 1, 4), "-", mo, "-", d),
```
[the remark this answers is not in the document]
Comment (bms63) [reply R/derive_date_vars.R]: This has been updated in the if/else statements

**Comment by bms63** [comment]
Hi Stefan @bundfussr 

Many thanks again with your help on this Review.  Learning a lot and while this one seemed like an easy issue to tackle, it has been quite a challenge for me!!  The adventures in cyclometric complexity continues!  

> According to our decision on Monday the `preserve` parameter should be added to functions which do imputation but do not provide a `date_imputation` parameter. This should affect `derive_param_tte()`, `derive_var_trtedtm()`, and `derive_var_trtsdtm()`.

Can we create a separate issue for these functions?  They seem a little more involved as I have been reviewing them and I would like to get this PR in before the release.  Happy to take them on after the release.  I'm just running out of time and like to wrap this up and help focus with release activities..

@thomas-neitmann I increased the cyclomatic complexity #560 from 15 to 20 in the `this branch` `.lintr `file.  Just wanted to give you a heads up!

**Comment by bundfussr** [comment]
@bms63 , I think the code produces correct results now. It could be simplified but we want to refactor it anyway. So we do not need to do it now.

As we are still in an experimental stage we can update the other functions in a separate issue. However we should not get used to this practice. Once we are in production we should not release half-implemented functionality.

Please update the news file. Then we can merge the PR.

**Comment by bms63** [comment]
@bundfussr Updated news.  Let me know if it makes sense.  Going to make separate issues for refactoring the code and updating This should affect `derive_param_tte()`, `derive_var_trtedtm()`, and `derive_var_trtsdtm()`.

File: NEWS.md [hunk 7]
```diff
@@ -32,10 +30,14 @@ are not recorded in the data. (#589)
 - The `derive_last_dose()` function has been deprecated in favor of
 `derive_var_last_dose_date()`. (#385)
 
+- Behavior of `derive_vars_dtm()` and `derive_vars_dt()` has changed with regards to `date_imputation` 
+when used in conjunction with new `preserve` argument. For example, if `date_imputation="MID"` then a date like  `2019---07` would become `2019-06-07` by setting `preserve` to `TRUE`.  However, if only using `date_imputation="MID"` with preserve set to `FALSE` or not invoked, then partial date, `2019---07`, will become `2019-06-15.` Similar behavior will happen for `"FIRST"` and `"LAST"` in `date_imputation` (#592).
```
Comment (bundfussr) [inline NEWS.md:34]: Sorry, but I think it does not really describe what has changed. I would change it to something like the following.
```
- The behavior of all functions providing the `date_imputation` parameter, e.g., `derive_vars_dtm()` and
`derive_vars_dt()` has changed for `date_imputation = "mid"`. Before the date was imputed as June 15th
if both month and day were missing. Now it is imputed as June 30th. For the old behavior please specify
`date_imputation = "06-15"`. Please note the the behavior has not changed if only the day is missing. In
this case the day is imputed as `15`.
```

**Comment by bms63** [comment]
@bundfussr Fixed up!

**Review by bundfussr — APPROVED** [review]

**Pull request merged**